lib/puppet/provider/mongodb.rb

```diff
@@ -132,9 +132,17 @@ def self.conn_string
 
   def self.db_ismaster
     cmd_ismaster = 'db.isMaster().ismaster'
-    cmd_ismaster = mongorc_file + cmd_ismaster if mongorc_file
     db = 'admin'
-    res = mongo_cmd(db, conn_string, cmd_ismaster).to_s.chomp
+
+    full_command = if mongorc_file
+                     mongorc_file + cmd_ismaster
+                   else
+                     cmd_ismaster
+                   end
+    res = mongo_cmd(db, conn_string, full_command).to_s.chomp
+    if res =~ %r{Authentication failed}
```

Comment:
You retry the same command twice if there's an authentication failure (previous iteration didn't have this). You probably want to change this to `if !mongorc_file && res =~ %r{Authentication failed}` to get the same result.